packages/server/src/shared/internal/serialize.ts

```diff
@@ -49,14 +50,16 @@ type SerializeObject<T extends object> = {
  * Example: { a: string | undefined} --> { a?: string}
  */
 
-type FilterDefinedKeys<TObj extends object> = {
-  [TKey in keyof TObj]: undefined extends TObj[TKey] ? never : TKey;
-}[keyof TObj];
+type FilterDefinedKeys<TObj extends object> = Exclude<
+  {
+    [TKey in keyof TObj]: undefined extends TObj[TKey] ? never : TKey;
+  }[keyof TObj],
+  undefined
+>;
 
-type UndefinedToOptional<T extends object> = {
+type UndefinedToOptional<T extends object> =
   // Property is not a union with `undefined`, keep as-is
-  [k in keyof Pick<T, FilterDefinedKeys<T>>]: T[k];
-} & {
-  // Property _is_ a union with `defined`. Set as optional (via `?`) and remove `undefined` from the union
-  [k in keyof Omit<T, FilterDefinedKeys<T>>]?: Exclude<T[k], undefined>;
-};
+  Pick<T, FilterDefinedKeys<T>> & {
+    // Property _is_ a union with `defined`. Set as optional (via `?`) and remove `undefined` from the union
+    [k in keyof Omit<T, FilterDefinedKeys<T>>]?: Exclude<T[k], undefined>;
+  };
```

Comment:
This looks quite "heavy" but before looked heavy too... reminds me of our `OmitNeverKeys` that ended up being slow.

FWIW, I think it's fine to expose never keys.... maybe an even better thing would be some DX like this:

```ts
const output = {
  bigint: 0n,
}

// ... then
serializedOutput.bigint
//                            ^? Error<"This type is not JSON-serializable">
```

Reply:
yeah this is the same as before, I just refactored it a bit. This part doesn't change with the `never` keys. It just has to be done unless we somehow come up with something more optimized.